HDR data is clipped in AVIF images

shallawa opened the pull request:
#### 2ba4ecb0408c03ab5abe7896df68492a784ebeff
<pre>
HDR data is clipped in AVIF images
<a href="https://bugs.webkit.org/show_bug.cgi?id=245858">https://bugs.webkit.org/show_bug.cgi?id=245858</a>
rdar://100879401

Reviewed by Simon Fraser.

This implements a temporary fix for rendering images with HDR data on macOS. The
system libraries on macOS BigSur and Ventura do a smooth conversion for the HDR
to SDR if the destination is a CGBitmapContext. This workaround will be used to
fix this problem by drawing the HDR image to a temporary local CGBitmapContext
before drawing it to the destination context.

* LayoutTests/platform/ios/fast/canvas/canvas-drawImage-hdr-video-expected.txt: Added.
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::draw): Deleted.
* Source/WebCore/platform/graphics/cairo/NativeImageCairo.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):

Canonical link: <a href="https://commits.webkit.org/260512@main">https://commits.webkit.org/260512@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4479903728437e90d1bdda7b8f6cdb218e002099

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17583 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41336 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117592 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/117798 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112373 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8861 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100713 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114254 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14266 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97488 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42222 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96244 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29129 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10395 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30477 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11151 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7390 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16541 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50073 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7266 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12738 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
```diff
@@ -69,6 +70,39 @@ DestinationColorSpace NativeImage::colorSpace() const
     return DestinationColorSpace(CGImageGetColorSpace(m_platformImage.get()));
 }
 
+static bool isHDRNativeImage(NativeImage& image)
```
`const NativeImage&` ?

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
```diff
@@ -69,6 +70,39 @@ DestinationColorSpace NativeImage::colorSpace() const
+        return;
+    }
+
+    // Create a temporary ImageBuffer for destinationRect with the current scaleFator.
```
The rest of this function is all about the HDR workaround, so I'd prefer the code be laid out to make that clearer (initially I missed the `! isHDRNativeImage()` early return and was confused). You could put the bulk of this function into a lambda, or another function, with a name that makes it clear that it's an HDR workaround.

There should be a FIXME comment referencing a radar that can tell us when we're able to remove the workaround.

shallawa replied:
A lambda was added because we need to call `GraphicsContext::drawNativeImageInternal()` which is private.

shallawa replied:
rdar://105525195 was filed and a FIXME comment was added.

smfr commented on Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
```diff
@@ -69,6 +70,39 @@ DestinationColorSpace NativeImage::colorSpace() const
+    }
+
+    // Create a temporary ImageBuffer for destinationRect with the current scaleFator.
+    auto imageBuffer = context.createScaledImageBuffer(destinationRect, context.scaleFactor(), DestinationColorSpace::SRGB(), RenderingMode::Unaccelerated, RenderingMethod::Local);
```
This buffer should either use the image's colorspace, or the destination colorspace, but not sRGB.

shallawa replied:
Fixed. DisplayP3 is used for iOS. Otherwise we match the destination context colorspace.

shallawa committed:
HDR data is clipped in AVIF images
https://bugs.webkit.org/show_bug.cgi?id=245858
rdar://100879401

Reviewed by Simon Fraser.

This implements a temporary fix for rendering images with HDR data on macOS. The
system libraries on macOS BigSur and Ventura do a smooth conversion for the HDR
to SDR if the destination is a CGBitmapContext. This workaround will be used to
fix this problem by drawing the HDR image to a temporary local CGBitmapContext
before drawing it to the destination context.

* LayoutTests/platform/ios/fast/canvas/canvas-drawImage-hdr-video-expected.txt: Added.
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::draw): Deleted.
* Source/WebCore/platform/graphics/cairo/NativeImageCairo.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):

Canonical link: https://commits.webkit.org/260512@main
